docs: add DID section in concepts (#82)
KILTprotocol/docs: docs/concepts/did.md

id: did
title: DIDs
---

import ThemedImage from '@theme/ThemedImage';

A KILT Decentralized Identifier (DID) is a string uniquely identifying each KILT user.
A DID can be thought of as a container of different keys that are all under the control of the same DID subject.
For the official W3C DID spec, please visit the [DID Core spec page](https://www.w3.org/TR/did-core/), while for the official KILT DID method specification, please visit the [KILT DID spec page](https://github.com/KILTprotocol/kilt-did-driver/blob/master/docs/did-spec/spec.md).

> The simplest type of KILT DID is a **light DID**, called this way because it can be generated and used offline without requiring any Internet connection (hence any connection with the KILT blockchain at all).

Reviewer: ```suggestion
The simplest type of KILT DID is a **light DID**, called this way because it can be generated and used offline without requiring any internet connection (hence any connection with the KILT blockchain at all).
```

Author: https://github.com/KILTprotocol/docs/pull/82/commits/27ff8acbe8ca3344d8a85d143427b5423803b27d

```

In the global URI space, Web3 names are prefixed with the `w3n:` URI namespace. 
For example, the full URI for the Web3 name `example-web3name` is `w3n:example-web3name`.

### Linking accounts to a Web3 name

Beyond linking a Web3 name, KILT allows DID owners to link multiple accounts to a single DID.
These accounts are not specific to the KILT blockchain but can reference any chain within the Dotsama ecosystem.
Each account <-> DID link requires the payment of a small deposit, which is returned whenever the link is removed.

For DIDs that have also claimed a Web3 name, the linking feature opens up the way to a host of possibilities, e.g., showing the Web3 name of a collator's account on the [KILT stakeboard](https://stakeboard.kilt.io/).

<ThemedImage
  alt="DID lookup diagram"
  sources={{
    light: '/img/concepts/did/did-lookup-light.png',
    dark: '/img/concepts/did/did-lookup-dark.png',
  }}
/>

:::caution
While multiple accounts can be linked to a DID, it is important to notice the difference between the two.

KILT *accounts* are classical blockchain accounts, that can be used to hold and send KILT tokens, as well as signing and submitting transactions.

> On the other hand, KILT *DIDs* are a L2 construct which are derived from KILT accounts, but are completely separated from them.

Reviewer 2: I'm not sure if L2 is understood widely.

```suggestion
On the other hand, KILT *DIDs* are a higher level construct which are derived from KILT accounts, but are completely separated from them.
```

Author: https://github.com/KILTprotocol/docs/pull/82/commits/68ce5d9232ea37164d25c622112f8ae659f344b7

This means that **KILT DIDs cannot hold any KILT tokens**.
DIDs are used to authorize (sign) some operations, but the resulting signature must then be submitted to the blockchain by a KILT account, which must pay for the transaction fees.

> Hence, even though a DID has the format `did:kilt:4rp4rcDHP71YrBNvDhcH5iRoM3YzVoQVnCZvQPwPom9bjo2e` with `4rp4rcDHP71YrBNvDhcH5iRoM3YzVoQVnCZvQPwPom9bjo2e` being a valid KILT account, there is no (immediate) relationship between the DID and this account and hence this account should not be considered in isolation as a KILT account, but only with its `did:kilt` prefix, which makes it a valid DID.

Reviewer 2: This was hard to understand for me

```suggestion
Hence, even though a DID has the format `did:kilt:4rp4rcDHP71YrBNvDhcH5iRoM3YzVoQVnCZvQPwPom9bjo2e` with `4rp4rcDHP71YrBNvDhcH5iRoM3YzVoQVnCZvQPwPom9bjo2e` being a valid KILT account, there is no (immediate) relationship between the DID and this account and hence this account should not be considered in isolation as a KILT account, but only with its `did:kilt` prefix, which makes it a valid DID.
```

Author: Do you think this is clearer now? https://github.com/KILTprotocol/docs/pull/82/commits/168243953993122eb20033e147ce65acc14c828c

Reviewer 2: Yes! very clear now. Thanks. :)

> Hence, if instructed to "*send some funds to the DID by using the account after the `did:kilt` prefix*", please ignore the advice, as sending funds to a DID can result in those funds being lost without the required technical expertise.

Reviewer 2: You used hence twice in a row. 
```suggestion
Therefor, if instructed to "*send some funds to the DID by using the account after the `did:kilt` prefix*", please ignore the advice, as sending funds to a DID can result in those funds being lost without the required technical expertise.
```

Author: https://github.com/KILTprotocol/docs/pull/82/commits/f88c41076ae7f68ec87b4c5bef28de5cbb6051b0

## DIDs and Web3 names

Web3 names are user-friendly aliases for KILT DIDs.
They serve the same purpose that domain names do for IP addresses: who knows the IP address under the `kilt.io` domain name? 🤷🏽‍♀️
There is a one-to-one relationship between DIDs and Web3 names.
This means that a KILT DID can be linked to at most one Web3 name, and a Web3 name can be claimed only by one DID.

Each Web3 name is globally unique within the KILT space, and is composed of a sequence of a minimum of 3 to a maximum of 32 characters taken from a specific character set to enhance human redability and reduce the chances of "two Web3 names looking the same, despite being different".
The character set include only:
- *lowercase letters*, from `a` to `z`
- *digits* from `0` to `9`
- the symbols `-` and `_`

A regex that would match all and only the allowed Web3 names would be the following:

```
^[a-z0-9\_\-]{3,32}$
```

In the global URI space, Web3 names are prefixed with the `w3n:` URI namespace. 
For example, the full URI for the Web3 name `example-web3name` is `w3n:example-web3name`.

> ### Linking accounts to a Web3 name

Reviewer 2: I would love to see a simple diagram for this. Just to make it easier.

Reviewer 3: Definitely! We need to step up our diagram game. If you want to, I can provide these @ntn-x2 .

Author: Phew! This cost me quite some time! Let me know what you guys think https://github.com/KILTprotocol/docs/pull/82/commits/f2ebb7497232d9fc2c92a65344f9722c4f7908ca.

Beyond linking a Web3 name, KILT allows DID owners to link multiple accounts to a single DID.

> These accounts are not specific to the KILT blockchain but can reference any chain within the Kusama ecosystem.

Reviewer 2: It's not only Kusama and it's rather limited by the used signing algorithms.

```suggestion
These accounts are not specific to the KILT blockchain but can reference any chain within the Kusama&Polkadot ecosystem that uses sr25519 or ed25519.
```

Author: What? We do support Ecdsa.

Reviewer 3: Maybe we want to say `Dotsama` instead?

Reviewer 2: > What? We do support Ecdsa.

Yeah forgot that. But the point is that it's not limited to kusama chains.

Author: https://github.com/KILTprotocol/docs/pull/82/commits/eaca2ed186e5dab481448978baf7acdc3bd4979e
